File: spring/poultry_hatchery_management_app/src/main/java/com/app/poultry_hatchery_management_app/service/HatchingIncubatorService.java
package com.app.poultry_hatchery_management_app.service;

import com.app.poultry_hatchery_management_app.dto.*;
import com.app.poultry_hatchery_management_app.model.*;
import com.app.poultry_hatchery_management_app.repository.HatchingIncubatorRepository;
import com.app.poultry_hatchery_management_app.repository.HatchingIncubatorSpaceRepository;
import com.app.poultry_hatchery_management_app.repository.HatchingTrolleyIncubatorSpaceAssignmentRepository;
import lombok.AllArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@AllArgsConstructor
@Service
public class HatchingIncubatorService {

    private final HatchingIncubatorRepository hatchingIncubatorRepository;
    private final HatchingIncubatorSpaceRepository hatchingIncubatorSpaceRepository;
    private final HatchingTrolleyIncubatorSpaceAssignmentRepository hatchingTrolleyIncubatorSpaceAssignmentRepository;

    private final HatchingTrolleyService hatchingTrolleyService;

    public List<HatchingIncubator> getAllHatchingIncubators() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null && authentication.getPrincipal() instanceof UserDetails) {
            User user = (User) authentication.getPrincipal();

            return hatchingIncubatorRepository.findAllByOrganisationId(user.getOrganisation().getId());
        }
        return null;
    }

    public Optional<HatchingIncubator> getHatchingIncubatorById(UUID incubatorId) {
        return hatchingIncubatorRepository.findById(incubatorId);
    }

    @Transactional(rollbackFor = Exception.class)
    public Optional<HatchingIncubator> postHatchingIncubator(PostHatchingIncubatorRequest request) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null && authentication.getPrincipal() instanceof UserDetails) {
            User user = (User) authentication.getPrincipal();

            HatchingIncubator incubator = HatchingIncubator.builder()
                    .organisation(user.getOrganisation())
                    .maxCapacity(request.maxCapacity())
                    .humanReadableId(request.humanReadableId())
                    .numberOfColumns(request.numberOfColumns())
                    .build();
            hatchingIncubatorRepository.save(incubator);

            for (int i = 0; i < request.maxCapacity(); i++) {
                PostHatchingIncubatorSpaceRequest spaceRequest = PostHatchingIncubatorSpaceRequest.builder()
                        .hatchingIncubatorId(incubator.getId())
                        .humanReadableId("S"+(i+1))
                        .build();
                Optional<HatchingIncubatorSpace> space = this.postHatchingIncubatorSpace(spaceRequest);
                if (space.isEmpty()) {
                    throw new RuntimeException("Error while creating hatching incubator space");
                }
            }
            return Optional.of(incubator);
        }
        return Optional.empty();
    }

    public Optional<HatchingIncubator> putHatchingIncubator(PutHatchingIncubatorRequest request) {
        Optional<HatchingIncubator> incubator = hatchingIncubatorRepository.findById(request.hatchingIncubatorId());
        if (incubator.isPresent()) {
            incubator.get().setMaxCapacity(request.maxCapacity());
            incubator.get().setHumanReadableId(request.humanReadableId());
            hatchingIncubatorRepository.save(incubator.get());

            return incubator;
        }
        return Optional.empty();
    }

    public void deleteHatchingIncubator(UUID hatchingIncubatorId) {
        hatchingIncubatorRepository.deleteById(hatchingIncubatorId);
    }

    public List<HatchingIncubatorSpace> getAllHatchingIncubatorsSpace(UUID hatchingIncubatorSpaceId) {
        return hatchingIncubatorSpaceRepository.findAllByHatchingIncubatorId(hatchingIncubatorSpaceId);
    }

    public Optional<HatchingIncubatorSpace> postHatchingIncubatorSpace(PostHatchingIncubatorSpaceRequest request) {
        Optional<HatchingIncubator> incubator = hatchingIncubatorRepository.findById(request.hatchingIncubatorId());
        if (incubator.isPresent()) {
            HatchingIncubatorSpace space = HatchingIncubatorSpace.builder()
                    .hatchingIncubator(incubator.get())
                    .humanReadableId(request.humanReadableId())
                    .isCurrentlyOccupied(false)
                    .build();
            hatchingIncubatorSpaceRepository.save(space);
            return Optional.of(space);
        }
        return Optional.empty();
    }

    public Optional<HatchingIncubatorSpace> putHatchingIncubatorSpace(PutHatchingIncubatorSpaceRequest request) {
        Optional<HatchingIncubatorSpace> space = hatchingIncubatorSpaceRepository.findById(request.hatchingIncubatorSpaceId());
        if (space.isPresent()) {
            space.get().setCurrentlyOccupied(request.isCurrentlyOccupied());
            space.get().setHumanReadableId(request.humanReadableId());
            hatchingIncubatorSpaceRepository.save(space.get());

            return space;
        }
        return Optional.empty();
    }

    public void deleteHatchingIncubatorSpace(UUID hatchingIncubatorSpaceId) {
        hatchingIncubatorSpaceRepository.deleteById(hatchingIncubatorSpaceId);
    }

    public List<HatchingTrolleyIncubatorSpaceAssignment> getAllTrolleysFromHatchingIncubator(UUID hatchingIncubatorId) {
        Optional<HatchingIncubator> incubator = hatchingIncubatorRepository.findById(hatchingIncubatorId);
        if (incubator.isPresent()) {
            return hatchingTrolleyIncubatorSpaceAssignmentRepository.findAllByHatchingIncubatorId(hatchingIncubatorId);
        }
        return List.of();
    }

    @Transactional
    public Optional<HatchingTrolleyIncubatorSpaceAssignment> postHatchingTrolleyToIncubatorSpace(PostHatchingTrolleyIncubatorSpaceAssignmentRequest request) {
        Optional<HatchingTrolley> trolley = hatchingTrolleyService.getHatchingTrolleyById(request.hatchingTrolleyId());
        Optional<HatchingIncubatorSpace> space = hatchingIncubatorSpaceRepository.findById(request.hatchingIncubatorSpaceId());

        if (trolley.isPresent() && space.isPresent()) {
            HatchingTrolleyIncubatorSpaceAssignment htisa = HatchingTrolleyIncubatorSpaceAssignment.builder()
                    .hatchingTrolley(trolley.get())
                    .hatchingIncubatorSpace(space.get())
                    .trolleyExitStamp(null)
                    .trolleyEntryStamp(LocalDateTime.now())
                    .build();
            hatchingTrolleyIncubatorSpaceAssignmentRepository.save(htisa);

            HatchingIncubatorSpace space2 = space.get();
            space2.setCurrentlyOccupied(true);
            hatchingIncubatorSpaceRepository.save(space2);

            return Optional.of(htisa);
        }
        return Optional.empty();
    }

    public Optional<HatchingTrolleyIncubatorSpaceAssignment> putHatchingTrolleyToIncubatorSpace(PutHatchingTrolleyIncubatorSpaceAssignmentRequest request) {
        Optional<HatchingTrolleyIncubatorSpaceAssignment> assignment =
                hatchingTrolleyIncubatorSpaceAssignmentRepository.findById(request.assignmentId());
        if (assignment.isPresent()) {
            if (assignment.get().getHatchingIncubatorSpace().isCurrentlyOccupied()) {
                assignment.get().getHatchingIncubatorSpace().setCurrentlyOccupied(false);
                assignment.get().setTrolleyExitStamp(LocalDateTime.now());
            } else {
                assignment.get().getHatchingIncubatorSpace().setCurrentlyOccupied(true);
                assignment.get().setTrolleyEntryStamp(LocalDateTime.now());
            }
            return assignment;
        }
        return Optional.empty();
    }

    @Transactional
    public void deleteHatchingTrolleyFromIncubatorSpace(UUID assignmentId) {
        hatchingTrolleyIncubatorSpaceAssignmentRepository.deleteById(assignmentId);
        Optional<HatchingTrolleyIncubatorSpaceAssignment> htisa =
                hatchingTrolleyIncubatorSpaceAssignmentRepository.findById(assignmentId);
        if (htisa.isPresent()) {
            HatchingIncubatorSpace space = htisa.get().getHatchingIncubatorSpace();
            space.setCurrentlyOccupied(false);
            hatchingIncubatorSpaceRepository.save(space);
            hatchingTrolleyIncubatorSpaceAssignmentRepository.delete(htisa.get());
        }
    }
}

File: angular/poultry_hatchery_management_app_angular/src/app/services/hatching-incubator/hatching-incubator.service.ts
import { Injectable } from '@angular/core';
import { HatchingIncubator } from '../../models/hatching-incubator.model';
import { HttpClient } from '@angular/common/http';
import { catchError, map, Observable, of, shareReplay } from 'rxjs';
import { apiUrl } from '../../app.config';
import { ApiPaths } from '../../api/api.paths';
import { AddressDetails } from '../../models/address-details.model';
import { OrganisationDetails } from '../../models/organisation-details.model';
import { PostIncubatorRequest } from '../../dto/post-incubator-request';
import { HatchingIncubatorSpace } from '../../models/hatching-incubator-space.model';
import { HatchingTrolleyIncubatorSpaceAssignment } from '../../models/hatching-trolley-incubator-space-assignment.model';
import { PostHatchingTrolleyToIncubatorSpaceRequest } from '../../dto/post-hatching-trolley-to-incubator-space-request';

@Injectable({
  providedIn: 'root'
})
export class HatchingIncubatorService {

  constructor(
    private http: HttpClient
  ) { }


  public getAllHatchingIncubators(): Observable<HatchingIncubator[]> {
    return this.http.get<any>(`${apiUrl}${ApiPaths.HatchingIncubatorPaths.GET_HATCHING_INCUBATOR}`).pipe(
      shareReplay(1),
      catchError(error => {
        return of([]);
      })
    );
  }

  public getHatchingIncubatorById(incubatorId: string): Observable<HatchingIncubator> {
    return this.http.get<HatchingIncubator>(`${apiUrl}${ApiPaths.HatchingIncubatorPaths.GET_HATCHING_INCUBATOR_BY_ID + incubatorId}`).pipe(
      shareReplay(1),
      catchError(error => {
        return of();
      })
    );
  }

  public postHatchingIncubator(body: PostIncubatorRequest): Observable<any> {
    return this.http.post<any>(`${apiUrl}${ApiPaths.HatchingIncubatorPaths.POST_HATCHING_INCUBATOR}`, body);
  }

  public getAllHatchingIncubatorSpaces(hatchingIncubatorId: string): Observable<HatchingIncubatorSpace[]> {
    return this.http.get<HatchingIncubatorSpace[]>(`${apiUrl + ApiPaths.HatchingIncubatorPaths.GET_HATCHING_INCUBATOR_SPACES + hatchingIncubatorId}`).pipe(
      catchError(error => {
        return of([]);
      })
    );
  }


  public getAllTrolleysCurrentlyInIncubator(incubatorId: string): Observable<HatchingTrolleyIncubatorSpaceAssignment[]> {
    return this.http.get<HatchingTrolleyIncubatorSpaceAssignment[]>(`${apiUrl + ApiPaths.HatchingIncubatorPaths.GET_HATCHING_TROLLEY_INCUBATOR_SPACE_ASSIGNMENT + incubatorId}`).pipe(
      catchError(error => {
        return of([]);
      })
    );
  }

  public deleteHatchingTrolleyFromIncubatorSpace(hatchingTrolleyIncubatorSpaceAssignmentId: string): Observable<any> {
    return this.http.delete<any>(`${apiUrl + ApiPaths.HatchingIncubatorPaths.DELETE_HATCHING_TROLLEY_FROM_INCUBATOR_SPACE + hatchingTrolleyIncubatorSpaceAssignmentId}`).pipe(
      catchError(error => {
        return of();
      })
    );
  }

  public postHatchingTrolleyToIncubatorSpace(body: PostHatchingTrolleyToIncubatorSpaceRequest): Observable<any> {
    return this.http.post<any>(`${apiUrl}${ApiPaths.HatchingIncubatorPaths.POST_HATCHING_TROLLEY_TO_INCUBATOR_SPACE}`, body).pipe(
      catchError(error => {
        return of();
      })
    );
  }
  
}
